classes/helper.php

```diff
@@ -34,18 +34,31 @@
 
 class helper {
 
-	static function purge($maxdays) {
-		global $CFG, $DB;
+	/**
+	 * Purges emails logged older than $maxdays
+	 * @param int $maxdays the number of days old a log should be before being purged
+	 * @param int $maxiterations the number of blocks of 1k records to delete as a fail safe against infinite loops.
+	 */
+	static function purge($maxdays, $maxiterations = 1000) {
+		global $DB;
 
-		$olderthantime = time() - $maxdays * 24 * 60 * 60;
+		$olderthantime = time() - $maxdays * DAYSECS;
 
-		// We can straight up delete records without attachments
-		$DB->delete_records_select('mail_log', 'timesent < ? AND attachment IS NULL', array($olderthantime));
+		// We can straight up delete records without attachments. This is very efficient.
+		$DB->delete_records_select('mail_log', 'timesent < ? AND attachment IS NULL', [$olderthantime]);
 
-		// If there's any left to delete, these must have attachments - they should be removed one-by-one
-		$logitems = $DB->get_records_select('mail_log', 'timesent < ?', array($olderthantime));
-		foreach ($logitems as $item) {
-			\local_maillog\helper::delete(array($item->id));
+		// The remaining have attachments which need to also be deleted.
+		// Delete these in chunks of 1k, as to not exceed the query parameter limit.
+		// Do this until there are no more records to delete (or ran out of iterations as a fail safe);
+		for ($i = 0; $i < $maxiterations; $i++) {
+			$records = $DB->get_records_select('mail_log', 'timesent < ?', [$olderthantime], '', 'id', 0, 1000);
+			$ids = array_column($records, 'id');
+
+			// Done - exit early.
+			if (empty($ids)) {
+				break;
```

Comment:
tiny tiny nitpick, I dont think this should be double indented.

Reply:
Its not - it just looks weird because the file is using tabs instead of 4 spaces. I had to turn off my auto tab to space to get it to match the rest of the file for consistency